### PR TITLE
Construct MulAddMul at gemm_wrapper! call sites

### LINKED ISSUE #684: mul! performance regression on master
On the current master branch, `mul!` of small matrices can be 10x slower than 1.3 (and also allocates)
```
julia> versioninfo()
Julia Version 1.4.0-DEV.556
Commit 4800158ef5* (2019-12-03 21:18 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin19.0.0)
  CPU: Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-8.0.1 (ORCJIT, skylake)
```

test code:

```
using LinearAlgebra
using BenchmarkTools

ndim = 3

m1 = rand(ComplexF64,ndim,ndim);
m2 = rand(ComplexF64,ndim,ndim);
ou = rand(ComplexF64,ndim,ndim);

@btime mul!($ou, $m1, $m2);
```

With the release-1.3 version,
```
33.471 ns (0 allocations: 0 bytes)
```
on master:
```
394.428 ns (1 allocation: 16 bytes)
```

While it's nano-seconds, when one has `mul!` in the inner-most/hot loop, this can easily translate to big performance degradation when most of the calculation is with such matrix productions. Larger matrices (e.g. `ndim = 30`) appears unaffected (dispatched to BLAS?) `2.732 μs (0 allocations: 0 bytes)` on 1.3 and `2.774 μs (0 allocations: 0 bytes)` on master.

### LINKED ISSUE #34013: mul! performance regression on master
On the current master branch, `mul!` of small matrices can be 10x slower than 1.3 (and also allocates)
```
julia> versioninfo()
Julia Version 1.4.0-DEV.556
Commit 4800158ef5* (2019-12-03 21:18 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin19.0.0)
  CPU: Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-8.0.1 (ORCJIT, skylake)
```

test code:

```
using LinearAlgebra
using BenchmarkTools

ndim = 3

m1 = rand(ComplexF64,ndim,ndim);
m2 = rand(ComplexF64,ndim,ndim);
ou = rand(ComplexF64,ndim,ndim);

@btime mul!($ou, $m1, $m2);
```

With the release-1.3 version,
```
33.471 ns (0 allocations: 0 bytes)
```
on master:
```
394.428 ns (1 allocation: 16 bytes)
```

While it's nano-seconds, when one has `mul!` in the inner-most/hot loop, this can easily translate to big performance degradation when most of the calculation is with such matrix productions. Larger matrices (e.g. `ndim = 30`) appears unaffected (dispatched to BLAS?) `2.732 μs (0 allocations: 0 bytes)` on 1.3 and `2.774 μs (0 allocations: 0 bytes)` on master.

### PR DESCRIPTION
**Pull request by tkf**
This is yet another PR to fix JuliaLang/LinearAlgebra.jl#684.  The idea is to move alpha and beta to the type domain as early as possible.  Maybe this is better than JuliaLang/julia#34384 as it relies less on inlining and constant propagation?

### CODE REVIEW

File: stdlib/LinearAlgebra/src/generic.jl [hunk 1]
```diff
@@ -25,10 +25,23 @@ struct MulAddMul{ais1, bis0, TA, TB}
...
+        if iszero(beta)
+            return MulAddMul{false,true,TA,TB}(alpha, beta)
+        else
+            return MulAddMul{false,false,TA,TB}(alpha, beta)
+        end
+    end
+end
 
-MulAddMul() = MulAddMul(true, false)
+MulAddMul() = MulAddMul{true,true}(true, false)
```
Comment (KristofferC) [inline stdlib/LinearAlgebra/src/generic.jl]:
```suggestion
MulAddMul() = MulAddMul{true,true,Bool,Bool}(true, false)
```

maybe?

**Comment by KristofferC** [comment]
Let's go with this one?

**Comment by tkf** [comment]
Yeah, I think this is better.  The latest commits in #34384 started breaking the tests.

**Comment by KristofferC** [comment]
Win32 CI hang happens on the other PRs as well.

**Comment by StefanKarpinski** [comment]
The part of https://github.com/JuliaLang/julia/pull/34384 that moves the calls to matmul2x2! and matmul3x3! before all the heavywight checking that they don't require seems like an independently good change, however, since it should reduce overhead in those cases. Separating higher overhead call to BLAS into its own subroutine also seems like a good thing. Not sure why this was suddenly decided to be much better.

**Comment by KristofferC** [comment]
Because it didn't rely on constant propagation, it seemed less brittle to compiler changes.

**Comment by StefanKarpinski** [comment]
Sure, that's good. The other parts of that change are good anyway though, without the parts that force inlining and constant propagation—i.e. that version of `gemm_wrapper!` should be easier for the compiler to deal with.

**Pull request merged**